### ISSUE TITLE
Concrete Syntax Model ready for bug reports?

### ISSUE THREAD

**Issue by cal101** [javaparser/javaparser#780]
Moin!

First let me thank you for the great work you are doing with java parser!

Seeing the concrete syntax commit I gave it a short try using 

                        final CompilationUnit cu = JavaParser.parse(content);
                        final String printed = ConcreteSyntaxModel.genericPrettyPrint(cu);
                        Files.write(out, printed.getBytes(StandardCharsets.UTF_8));
                        if (!content.equals(printed)) {
                            fails.add(path.toString());
                            //System.out.println("diff failed: " + path);
                        }

And trying it on a bunch of source files.

Not really successful ;-)

java.lang.UnsupportedOperationException: MemberValuePair
	at com.github.javaparser.printer.ConcreteSyntaxModel.forClass(ConcreteSyntaxModel.java:766)

java.lang.UnsupportedOperationException: SingleMemberAnnotationExpr
	at com.github.javaparser.printer.ConcreteSyntaxModel.forClass(ConcreteSyntaxModel.java:766)

That leads to my questions:

Did I used it wrong?
Is it worth adding concrete bug reports and extract samples or should I wait for the dust to settle?
Where parsing and printing was OK I see indentation errors especially on comments and annotations.
Or is it something I can hack on myself? (A short hint HOW or WHERE to tackle that would be nice.)

Cal

**Comment by ftomassetti** [comment]
The Concrete Syntax Model is very, very new so and there are a few kinds of nodes which could not be covered yet. A few are going to be fixed as part of #654 but for others a PR would be greatly appreciated :D 

Reporting issue is **ALWAYS** a good thing and extremely useful, so please feel free do it!

**Comment by ftomassetti** [comment]
About changing the behavior of nodes which are handled by the CSM: you can just look into ConcreteSyntaxModel. For each Node class there is a CsmElement associated (typically a CsmSequence). You should be able to fix every possible issue changing the specific CsmElement used for the Node class. Documentation on how to do so will follow, for now I would just look at other CsmElement usages and to the corresponding method of the PrettyPrintVisitor. In most cases the CsmElement is just a translation of what was present in the PrettyPrintVisitor.

Feel free to ask as many questions as you need.

**Comment by cal101** [comment]
OK, cool.

SingleMemberAnnotationExpr and MemberValuePair were easy to add.
The missing AnnotationDeclaration is too big to digest for me.

How do I change IMPORTS into preserving newlines between single IMPORT?
How to "unparse" : ("\n" | "\SPACE) *

How to model optional leading or trailing whitespace?

**Comment by ftomassetti** [comment]
Perfect because AnnotationDeclaration and AnnotationMemberDeclaration are implemented in #654 

About preserving newlines about imports: this is not going to work as you expect. 

The ConcreteSyntaxModel describe how each node should look like. This model is then used to pretty-print the AST (this code is present in master) and it will be used also to print considering lexical preservation (this code is already in #654, waiting to be cleaned up and commit). So with the code in master you can just pretty print, if you have 3 white lines between your imports they will just be not printed. Not until we get #654 in.

There is so no need to model explicitly optional whitespace because the LexicalPreservingPrinter take care of it.

**Comment by cal101** [comment]
OK, I think I understood.

Is your current open pull request accessible somehow for me to play with? 
I know it's work in progress and it's RAW.
Guess I am lacking some github knowhow.

**Comment by ftomassetti** [comment]
Yes, you can look into it and play with it, you can even send me a PR for me to include your changes on that branch, however this requires some GitHub expertise :)

**Comment by cal101** [comment]
Grrr... I forgot to look into the branches of your fork. Found it.

**Comment by cal101** [comment]
In case you didn't saw it, the test cases may be useful:
    https://github.com/ftomassetti/javaparser/issues/4

**Comment by matozoid** [comment]
@ftomassetti the CSM has a label now :)

**Comment by ftomassetti** [comment]
It is a big day for the CSM :)

**Comment by cal101** [comment]
Great work, please keep on going!
CSM and Lexical Preservation is a big step forward!

**Comment by ftomassetti** [comment]
thank you @cal101 , your feedback has been invaluable. Many of the issues you have reported and the solutions you suggested are already included in #654 while the remaining ones will be addresses later.

Thanks a million!

**Comment by cal101** [comment]
A pleasure to offer my EUR 0,02.
I saw the fixes already because I followed your changes and forked your repo to test "bleeding edge".
Your history rewriting made my life interesting but I learned some more git tricks to deal with that. ;-)
Do NOT change anything of how you are doing things because of me forking  YOUR repo.
I expected those problems and *I* deal with that.
I tested the preservation on about 70k java files and have only 8 problems left that I have to analyze (this uses my unicode POC code based on current issue124 branch).
Keep on going!
